Adding in the ability to deal with multi-core

tspayne87 opened the pull request:
This is an attempt to deal with multi-core problem
Fixes #421

gjunge commented:
@tspayne87 thanks for the PR! I'm currently on holidays, I will review it at the end of the month.

gjunge commented on Microsoft.DependencyInjection.SolrNet/ServiceCollectionExtensions.cs:
```diff
@@ -86,5 +88,19 @@ private static IServiceCollection AddSolrNet(this IServiceCollection services, I
+        /// <param name="url">The url for the second core.</param>
+        /// <typeparam name="TModel">The type of model that should be used for this core.</typeparam>
+        /// <returns>The dependency injection service.</returns>
+        public static IServiceCollection AddSolrCore<TModel>(this IServiceCollection services, string url)
```
I'm a little bit on the fence regarding the naming. 
It's a little confusing that:
1) I first have to call `AddSolrNet` (and already pass in a URL, but no type)
2) Then for the next core call `AddSolrCore<> ()` and pass in another URL with a type.

Perhaps we could go the same way as DbContext:
``` csharp
services.AddDbContext<SomeContext>(options => someoptions) ;
services.AddDbContext<OtherContext>(options => someotheroptions);
```

So in our case:
Legacy:
`public static IServiceCollection AddSolrNet(this IServiceCollection services, string url)`
New:
` public static IServiceCollection AddSolrNet<TModel>(this IServiceCollection services, string url)`

I would like to see that calling the new method, will also work without the need to call the legacy one. 
So we should take the 'general DI setup' out of the legacy method, and into a private helper, and make sure inside that helper method that it does its work only once.

So the goal is that in my `startup.cs` I would be able to write:

services.AddSolrNet<Foobar>("http://solr.com/foobar");
services.AddSolrNet<Testing>("http://solr.com/testing");

tspayne87 commented:
Added in suggested change now you would do something like this.

```csharp
services.AddSolrNet("http://solr.com/foobar");
services.AddSolrNet<TestingModel>("http://solr.com/testing");
```

However, calling the AddSolrNet without a generic class for a second time would result in nothing being changed or added.  Should I just remove the AddSolrNet option or just allow for that to happen and just put a note in for that.

gjunge commented:
@tspayne87 could perhaps throw an `InvalidOperationException` in that case

tspayne87 commented:
That would be a good idea, you got a specific message in mind?

gjunge commented:
Something like:

> Only one non-typed Solr Core can be registered. In order to use multiple cores, use the typed AddSolrNet<>() overload.

Although `non-typed` is not the best wording, thought before about `generic` but that is confusing. Open to suggestions though, and I think that whatever we type will already help to user understand what is wrong.

tspayne87 commented:
Added in the exception operations, also I have been thinking about the class ```SolrCore```.  Since we are breaking the cores up into separate overloaded methods I do no think that it is needed anymore since we are just using it to store the url and nothing else.  Just wanted to ask before removing the code.

gjunge commented:
Yeah, thats a good idea.

tspayne87 commented:
I went ahead and removed the SolrCore class and added in some more documentation to the methods.  Was there anything else you can think of that would be good to add or remove?

gjunge commented:
If it wouldn't be too much of a hassle , could you add an example of the new usage to here:
https://github.com/SolrNet/SolrNet/blob/master/Documentation/Initialization.md#microsoftextensionsdependencyinjection-aspnet-core 
(Documentation/Initialization.md)

And here:
https://github.com/SolrNet/SolrNet/blob/master/Documentation/Multi-core-instance.md

That would really advertise your solution to the users. 

Thanks a million!

tspayne87 commented:
Added in some documentation to the two files you suggested, sorry it took so long been bogged down pretty hard with work.

gjunge commented:
@tspayne87  Thanks a million!
